src/train.js

```diff
@@ -141,50 +145,59 @@ async function trainOnTabs() {
     // Grey out Train button:
     const trainButton = document.getElementById('train');
     trainButton.disabled = true;
-    const progressBar = document.getElementById('progress');
-    progressBar.setAttribute('style', '');
 
-    // TODO: Using "active" here rather than a tab ID presents a race condition
-    // if you quickly switch away from the tab after clicking the Train button.
-    const tabs = (await browser.tabs.query({currentWindow: true, active: false}));
-    await setViewportSize(tabs[0], 1024, 768);  // for consistent element sizing in samples due to text wrap, etc.
+    // Show progress bar and output.
+    g_progressBar.classList.remove('hidden');
+    document.getElementById('output').classList.remove('hidden');
+
+    try {
 
-    const rulesetName = document.getElementById('ruleset').value;
-    const tuner = new Tuner(tabs, rulesetName);
-    const [tunedCoeffs, tunedCost] = await tuner.anneal(updateProgress);
+        // TODO: Using "active" here rather than a tab ID presents a race condition
+        // if you quickly switch away from the tab after clicking the Train button.
+        const tabs = (await browser.tabs.query({currentWindow: true, active: false}));
+        await setViewportSize(tabs[0], 1024, 768);  // for consistent element sizing in samples due to text wrap, etc.
 
-    progressBar.setAttribute('style', 'display: none');
-    progressBar.setAttribute('value', 0);
-    trainButton.disabled = false;
+        const rulesetName = document.getElementById('ruleset').value;
+        const tuner = new Tuner(tabs, rulesetName);
+        await tuner.anneal(updateProgress);
+
+    } finally {
+        // Restore UI state, leaving output visible.
+        g_progressBar.classList.add('hidden');
+        g_progressBar.setAttribute('value', 0);
+        trainButton.disabled = false;
+    }
 }
 
 function empty(element) {
-    while (element.firstChild) {
-        element.removeChild(element.firstChild);
+    while (element.lastChild) {
```

Comment:
Just so it doesn't look as goofy if it shows while it happens? Or does it prevent reflows or something?

Reply:
i added this when `empty` was being called a lot; i read somewhere that it's faster, with data to back it up.

however your comment made me wonder if this was still the case and it doesn't appear to be (i suspect if we were removing an element that wasn't first or last it would be a different story).  will revert.

Reply:
Interesting historical note, nonetheless!